Clang cevian

cevian opened the pull request.

Ngalstyan4 committed:
Implement show_chunks in C and have drop_chunks use it

Timescale provides an efficient and easy to use api to drop individual
chunks from timescale database through drop_chunks. This PR builds on
that functionality and through a new show_chunks function gives the
opportunity to see the chunks that would be dropped if drop_chunks was run.
Additionally, it adds a newer_than option to drop_chunks (also supported
by show_chunks) that allows to see/drop chunks in an interval or newer
than a point in time.

This commit includes:
    - Implementation of show_chunks in C
    - Additional helper functions to work with chunks
    - New version of drop_chunks in sql that uses show_chunks. This
      	  also adds a newer_than option to drop_chunks
    - More enhanced tests of drop_chunks and new tests for show_chunks

Among other reasons, show_chunks was implemented in C in order
to be able to have both older_than and newer_than arguments be null. This
was not possible in SQL because the arguments had to have polymorphic types
and whether they are used in function body or not, PL/pgSQL requires these
arguments to typecheck.

erimatnor committed:
Add support for functions on open (time) dimensions

TimescaleDB has always supported functions on closed (space)
dimension, i.e., for hash partitioning. However, functions have not
been supported on open (time) dimensions, instead requiring columns to
have a supported time type (e.g, integer or timestamp). This restricts
the tables that can be time partitioned. Tables with custom "time"
types, which can be transformed by a function expression into a
supported time type, are not supported.

This change generalizes partitioning so that both open and closed
dimensions can have an associated partitioning function that
calculates a dimensional value. Fortunately, since we already support
functions on closed dimensions, the changes necessary to support this
on any dimension are minimal. Thus, open dimensions now support an
(optional) partitioning function that transforms the input type to a
supported time type (e.g., integer or timestamp type). Any indexes on
such dimensional columns become expression indexes.

Tests have been added for chunk expansion and the hashagg and sort
transform optimizations on tables that are using a time partitioning
function.

Currently, not all of these optimizations are well supported, but this
could potentially be fixed in the future.

davidkohn88 committed:
Refactor compatibility functions and code to support PG11

Introduce PG11 support by introducing compatibility functions for
any whose signatures have changed in PG11. Additionally, refactor
the structure of the compatibility functions found in compat.h by
breaking them out by function (or small set of similar functions)
so that it is easier to see what changed between versions and maintain
changes as more versions are supported.

In general, the philosophy has been to try for forward compatibility
wherever possible, so that we use the latest versions of function interfaces
where we can or where reasonably convenient and mimic the behavior
in older versions as much as possible.

JLockerman committed:
Initial Timescale-Licensed-Module and License-Key Implementation

This commit adds support for dynamically loaded submodules to timescaledb
as well an initial license-key implementation in the tsl subdirectory.
Dynamically loaded modules allow our users to determine which licenses they
wish to use for their version of timescaledb; if they wish to only use
Apache-Licensed code, they do not load the Timescale-Licensed submodule. Calls
from the Apache-Licensed code into the Timescale-Licensed submodule are
handled via dynamicaly-set function pointers; see tsl/src/Readme.module.md for
more details.

This commit also adds code for license keys for the ApacheOnly, Community, and
Enterprise editions. The license key determines which features are enabled,
and controls loading the submodule: when a license key that requires the
sub-module is installed, the module is automatically loaded.
Currently the ApacheOnly and Community license-keys are hardcoded to be
"ApacheOnly" and "Community" respectively. The first version of the enterprise
license-key is described in tsl/src/Readme.module.md

Commit:
Add logic for automatic DB maintenance functions

This commit adds logic for manipulating internal metadata tables used for enabling users to schedule automatic drop_chunks and recluster policies. This commit includes:

- SQL for creating policy tables and chunk stats table
- Catalog code and C code for accessing these three tables programatically
- Implement and expose new user API functions:  add_*_policy and remove_*_policy
- Stub scheduler logic for running the policies

svenklemm committed:
Move int time_bucket functions with offset to C

svenklemm committed:
Check extension exists for DROP OWNED and DROP EXTENSION

In a database without timescaledb installed drop owned
failed with an error while trying to lookup the extension
owner.

Commit:
Provide helper function creating struct from tuple

Refactored the boilerplate that allocates and copies over data from a tuple to a struct. This is typically used in the scanner context in order to read rows from a SQL table in C.

niksajakovljevic committed:
Optimize FIRST/LAST aggregate functions

If possible replace aggregate functions FIRST/LAST with subqueries of the form
(SELECT value FROM table WHERE sort IS NOT NULL AND existing-quals ORDER BY sort ASC/DESC
LIMIT 1).
Given a suitable index on sort column, this plan can be much faster then scanning all the
rows and running an aggregate function.
The optimization can't be performed if:
- query uses GROUP BY or WINDOW function
- query contains CTEs
- query contains other aggregate functions (eg. Combining MIN/MAX with FIRST/LAST. We can't
	optimize accross different aggregate functions)
- query uses JOIN
- FIRST/LAST used in ORDER BY

Optimization also works with subqueries, or if FIRST/LAST is used in CTE subquery.

In order to standardize existing FIRST/LAST aggregate function with PostgreSQL and
FIRST/LAST optimization, we exclude NULL values in sort by column.

Commit:
Ensure deterministic results in cluster test

Adding ORDER BY to some select statements in cluster.sql test to ensure determinism.

cevian committed:
Add scripts for clang-format

We are switching from pgindent to clang-format for several
reasons:
1) pgindent is harder to set up and depends on other tools
like gobjdump that differ between OSes making it harder
to set up a proper development environment.
2) clang-format is more standard and works better
with various IDEs and editors.

Our clang-format wrapper script is transactional in that this
version of the script copies all files to a temporary directory,
runs clang-format in the temp dir, and only then copies the files
back. This should prevent mis-formatting when the script is cancled
in the middle of running.

We changes travis to use clang-format instead of pgindent
and thus it uses a different docker image to do the format test.
We also changes travis to use the new clang docker
image for the license check and prefix checks for consistency.
This also paves the way for us to use more clang features for tests
in the future.

Commit:
Change to heap allocation for runtime array size

Surprisingly, the compiler didn't complain about allocating a stack array with a size only known at runtime. Modified memory allocation to be heap-allocation to remedy this error.

niksajakovljevic committed:
Exclude unneeded chunks for IN/ANY/ALL operators

If IN/ANY/ALL operator is used with explicit values
 we can effectively restrict chunks that need to be scanned.
Here are some examples of supported queries:
- SELECT * FROM hyper_with_space_dim WHERE time < 10 AND device_id IN ('dev5','dev6','dev7','dev8');
- SELECT * FROM hyper_with_space_dim WHERE device_id = ANY(ARRAY['dev5','dev6']) AND device_id = ANY(ARRAY['dev6','dev7']);

There are som cases that are not optimized:
- subqueries within IN/ANY/ALL
- open dimension (eg. time)  when using IN/ANY with multiple args
- NOT operator

did-g committed:
set client_min_messages to 'error' not 'fatal'

after postgres commit c09daa9104099422ea998e0398934ca82eb37898 'fatal' is either not valid
or replaced by 'error'.

JLockerman committed:
Fix version_parse on windows

snprint has different behavior on some windows versions than on unix.
This commit replaces the snprintf version with one that uses sscanf,
which behaves the same on all platforms. (This is a higher priority
commit than it would appear, since we're tracking down a stack-overflow,
and this is one of the candidates f

erimatnor commented on src/agg_bookend.c:
```diff
@@ -34,24 +34,23 @@ TS_FUNCTION_INFO_V1(ts_bookend_deserializefunc);
-	Oid			type_oid;
-	bool		is_null;
-	Datum		datum;
+	Oid type_oid;
```
This still looks different. Spaces vs tabs?

cevian replied:
The difference here is that fields are not aligned instead one space is entered after a field. Clang does not support aligned declarations with the point right format. So we'd have to have unaligned fields. I think this is fine but if you disagree we can't use clang

cevian replied:
The difference here is that fields are not aligned instead one space is entered after a field. Clang does not support aligned declarations with the point right format. So we'd have to have unaligned fields. I think this is fine but if you disagree we can't use clang

erimatnor commented on src/agg_bookend.c:
```diff
@@ -395,12 +391,12 @@ ts_first_sfunc(PG_FUNCTION_ARGS)
 /* last(internal internal_state, anyelement value, "any" comparison_element) */
-Datum
-ts_last_sfunc(PG_FUNCTION_ARGS)
+Datum ts_last_sfunc(PG_FUNCTION_ARGS)
```
The `Datum` should be on its own line here.

cevian replied:
I don't think clang supports this

cevian replied:
I don't think clang supports this

erimatnor reviewed (commented):
This looks like it is still changing the formatting, especially it looks like it is not doing tab spacing correctly. It also doesn't properly put a function's return type on its own line.

JLockerman committed:
Get the hostname of the server from env instead of hardcoding localhost

fixes regresschecklocal for remote server

JLockerman committed:
Add appveyor.yml

After this commit AppVeyor will run all TimescaleDB tests on windows

JLockerman committed:
Fix casing in SQL license header to be consistent with elsewhere

erimatnor commented on .clang-format:
```diff
@@ -0,0 +1,119 @@
+JavaScriptQuotes: Double # N/A js
+JavaScriptWrapImports: true # N/A js
+KeepEmptyLinesAtTheStartOfBlocks: false
+MacroBlockBegin: '' # regex of macros that behave like '{'
```
Shouldn't we put things like `PG_TRY()` here? Or does this not apply because it has braces?

cevian replied:
Does not apply because of braces. bgw/job.c has a block and it looks fine.

cevian committed:
Fix upsert TLE translation when mapping variable numbers

Previously, there was a bug when chunks contained dropped columns
in the mapping of variable numbers. This PR fixes the bug and adds a
test for this case.

Commit:
Remove version-checking from client-side

Because the server now sends a boolean indicating whether the local version is up-to-date, we simply validate the server's response and remove any version-checking code from the client-side. The client now only checks for well-formed versions before printing, if the local version is not up-to-date.

svenklemm committed:
Use PG_INTXX_MAX macros instead of INTXX_MAX

Commit:
Clean up catalog code

Organize catalog.c so that functions that access the Catalog struct are physically separate from the functions that modify the actual catalog tables on disk. Also added macros and made static some functions that aren't used outside the catalog files. Also refactored out the CatalogDatabaseInfo struct, which is independent of the Catalog struct and will be reused in the future.

JLockerman committed:
allow loader tests to run on windows

svenklemm committed:
Add view to show hypertable information

Add timescaledb_information.hypertable view that lists
hypertables, their owner, number of chunks and storage size

cevian committed:
Backpatch TupleDescAttr for 9.6.3 and 9.6.4

svenklemm committed:
Change size utility and job functions to STRICT

Change hypertable_relation_size, hypertable_relation_size_pretty,
chunk_relation_size, chunk_relation_size_pretty, indexes_relation_size,
indexes_relation_size_pretty, partitioning_column_to_pretty,
insert_job and delete_job to STRICT

svenklemm committed:
Add checks for NULL arguments to DDL functions

Add checks for NULL arguments to ts_chunk_adaptive_set,
ts_dimension_set_num_slices, ts_dimension_set_interval,
ts_dimension_add and ts_hypertable_create

svenklemm committed:
Add tests for calling functions with NULL arguments

Ngalstyan4 committed:
Fix Datum conversion issues

svenklemm committed:
Set ORS to empty string instead of NUL in license check script

This fixes multiple warnings in bash 4.4:
scripts/check_file_license_apache.sh: line 22: warning: command substitution: ignored null byte in input
and is also more correct since we dont want to inject 0-Bytes into the string

did-g committed:
Ensure deterministic results in bw scheduler test

Commit:
Implement drop_chunks in C

Remove the existing PLPGSQL function that implements drop_chunks, replacing it with a direct call to the C function, which also implements the old PLPGSQL checks in C. Refactor out much of the code shared between the C implementations of show_chunks and drop_chunks.

svenklemm committed:
Refactor ChunkDispatch to remove ChunkDispatchInfo

svenklemm committed:
Remove unused functions from utils.c

Remove int_cmp, create_fmgr and makeRangeVarFromRelid from utils.c
since they were not used and had no test coverage.

svenklemm committed:
Remove EXEC_BACKEND tests from travis

Commit:
Remove unused function from hypertable.c

hypertable_delete_by_id has long since been replaced by hypertable_delete_name and isn't used anywhere in the codebase.

mfreed committed:
Fix header formatting in markdown subtitle

Commit:
Move BGW job tester functions to test module

Previously, delete/insert_job were exposed, even though they were only used for testing. Moved both the C implementations and the SQL functions to the test module.

Commit:
Make views test more deterministic

JLockerman committed:
Release 1.0.1 on master, prepare for 1.1.0-dev

JLockerman committed:
Add ts_ prefix to everything in headers

Future proofing: if we ever want to make our functions available  to
others they’d need to be prefixed to prevent name collisions. In
order to avoid having some functions with the ts_ prefix and
others without, we’re adding the prefix to all non-static
functions now.

did-g committed:
Only set soname if it's used

inval_cache_callback call recursively do_load but only one has loaded == false
after calling extension_version().

did-g committed:
Plug a trivial memory leak

erimatnor commented on src/bgw/job.c:
```diff
@@ -248,25 +263,23 @@ bgw_job_tuple_delete(TupleInfo *ti, void *data)
+	Catalog *catalog = catalog_get();
+
+	ScannerCtx scanctx = { .table = catalog->tables[BGW_JOB].id,
+						   .index = CATALOG_INDEX(catalog, BGW_JOB, BGW_JOB_PKEY_IDX),
```
This looks pretty terrible now with Clang. Can we do something to fix it?

cevian replied:
Fixed. Designated initializers need a final comma at the last element, then clang-format fixes this.

erimatnor commented on src/bgw/job.c:
```diff
@@ -21,12 +21,11 @@
-	[JOB_TYPE_VERSION_CHECK] = "telemetry_and_version_check_if_enabled",
-	[JOB_TYPE_UNKNOWN] = "unknown"
-};
+const char *job_type_names[_MAX_JOB_TYPE] = { [JOB_TYPE_VERSION_CHECK] =
```
Here it doesn't line-break the first member, which leads to weird formatting.

cevian replied:
fixed with final comma

erimatnor commented on scripts/clang_format_wrapper.sh:
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
```
Add comment/documentation of why we have this script and what it is doing

cevian replied:
fixed

cevian committed:
Add final commas to designated initializers

Final commas are needed for clang-format to format the
code appropriately.

JLockerman committed:
Add more license key fields and telemetry

Add more license fields and useful for tracking which kinds of licenses
are in use.

niksajakovljevic committed:
Exclude paths where folder/name starts with '.'. Previous regex would exclude everything if any of the folders in your path contains a dot (eg. /github.com/

amytai committed:
Recluster and drop chunks scheduling code

For both recluster and drop_chunks policies, adds the scheduler code to run the appropriate function after reading the policy arguments from the appropriate internal policy table. For the recluster jobs, also picks chunks that are eligible for recluster, based on the current selection: 1) chunks at least 3rd newest chunk, 2) have not been reclustered before. The logic will try to pick the oldest such chunk.

JLockerman committed:
Check that functions start with the ts_ prefix in travis

Ngalstyan4 committed:
Run custom_type tests only on 64-bit architectures

svenklemm committed:
Fix interval_from_now_to_internal timestamptz handling

fix interval_from_now_to_internal to handle timezone properly
for timestamptz and simplify code

JLockerman committed:
Add the ability to ignore tests from the command line using IGNORES

JLockerman committed:
Have AppVeyor ignore the bgw_db_scheduler test for now

svenklemm committed:
Set C standard to C11

amytai committed:
Update licensecheck to look for correct string

Previously, files under tsl/test/ were expected to have "Apache" rather than "Timescale" license in their license header. This has been fixed.

svenklemm committed:
Remove unused functions from trigger.c

svenklemm committed:
Remove unused functions from utils.c

Remove the following unused functions:
_timescaledb_internal.to_microseconds(TIMESTAMPTZ)
_timescaledb_internal.to_timestamp_pg(BIGINT)
_timescaledb_internal.time_to_internal(anyelement)

svenklemm committed:
Remove compat.c

The only function in compat.c was oid_cmp which had no users.

svenklemm committed:
Fall back to commit hash in update tests

Travis CI occasionally has trouble figuring out the branch, so this
change makes sure it falls back to using at least a commit hash

amytai committed:
Add alter_policy_schedule API function

JLockerman committed:
Have prefix check verify all functions in headers and clean up missings

This commit makes four changes:
1. Adds the ts_ prefix to some functions that were missed in the last
   pass
2. Adds static to some variables that were confined to a single file but
   missing it
3. Adds the ability to disable using "hidden" visibility by default by
   setting the USE_DEFAULT_VISIBILITY variable
4. Switches the prefix-check in travis to use the flag defined in 3 so
   that the checker now checks all non-static symbols we define

erimatnor committed:
Use create_upper_paths_hook for hypertable insert plan modifications

This change refactors how we do planner modifications to insert data
into a hypertable. Instead of altering the finished plan tree produced
by the standard planner, we now use the create_upper_paths_hook to do
these modifications at the path stage during planning.

Apart from being a cleaner, and arguably more principled, way of doing
plan modifications, it also gives us better ability to modify plan
state during planning since the hook happens during the normal planner
run and not after. For instance, our CustomScan insert plan is now
created using a callback that has access to the PlannerInfo data,
which will be useful when planning, e.g., foreign queries using the
foreign data wrapper API.

amytai committed:
Fix PG11 compatibility for current tsl/ module

JLockerman committed:
Release 1.1.0

PG11!
Otherwise, mainly performance improvements and bugfixes.

JLockerman committed:
Prepare the repo for the 1.2.0 development cycle

JLockerman committed:
Fix accidental usage of c99 for loop

mfreed committed:
Add PostgreSQL License information to NOTICE and selected files

JLockerman committed:
Free ChunkInsertStates when the es_per_tuple_exprcontext is freed

Instead of freeing ChunkInsertStates immediately when flushed from the
SubspaceStore, defer the freeing until the es_per_tuple_exprcontext is
being cleaned up. This ensures that pointers into the CIS don't become
invalidated without us freeing any of postgres's memory, like we were
doing in the prior fix; this fixes a segfault in COPY.

In the normal INSERT case these frees happen at approximately the same
time so there should be no regression in memory usage, while if there
was a delay in freeing the es_per_tuple_exprcontext our old strategy of
clearing the context ourselves was invalid.

svenklemm committed:
Add gapfill query support

This patch adds first level support for gap fill queries, including
support for LOCF (last observation carried forward) and interpolation, without
requiring to join against `generate_series`. This makes it easier to join
timeseries with different or irregular sampling intervals.

svenklemm committed:
Remove unused functions

Remove the following unused functions:
ts_cache_switch_to_memory_context
ts_chunk_free
ts_chunk_exists
ts_chunk_index_delete_children_of
ts_chunk_index_delete_by_hypertable_id
ts_hypertable_scan_relid
ts_tablespaces_clear
ts_tablespaces_delete

svenklemm committed:
Fix ORS in check_file_license.sh

svenklemm committed:
Fix return value of gapfill_aggref_mutator

Ngalstyan4 committed:
Fix string conversion format in timer_mock

davidkohn88 committed:
Avoid loading twice in parallel workers and load only from $libdir

We removed a parallel workers check a while back because there were problems
with the macro on Windows, we define a new macro that works on Windows as we
received reports of some instances in which the Timescale library appeared to
be loading twice in parallel workers. Parallel workers instead restore the
library state of their parent process via their normal pathway and the loader
will take no action.

Additionally, prepend "$libdir/" to all calls to .so's so that we don't end up
accidentally loading from a place different from where we've loaded when we
create functions, which all refer to $libdir. This would only affect users
who have multiple separate .so's in different directories, but could be another
case where multiple .so's were loaded.

RobAtticus committed:
Update README with pointers to tools and fixes

The installation links were a bit out-of-date given our new page
structure. Also, our tools were not mentioned anywhere, so now the
README has pointers to them at the bottom, and in particular a
brief section to point to timescaledb-tune.

mfreed committed:
Add Timescale License (TSL)

dianasaur323 committed:
Update README with new license

amytai committed:
Add enterprise checks to the 5 enterprise-only functions

davidkohn88 committed:
Don't throw errors when extension is loaded but not installed yet

Pg_upgrade was failing in extension_version() as it tries to check
that libraries that have functions associated with them are
available before copying data from the old cluster over to the new.
(Therefore the extension does not yet exist as the data has not been
copied and the check for extension_version fails, but it should not
be called if the extension does not yet exist).

erimatnor committed:
Do not exit immediately on failure in update test script

This fixes a number of issues with the update test script that
prohibited it from displaying a log of any failing tests. In
particular, the script needs to be run with `set +e` to ensure a test
error doesn't immediately cause an exit without recording the failed
log.

JLockerman committed:
Update testrunner to only start if there are tests to run

Exactly like it says, for instance it you run with
TESTS=append the isolation test runner won't even start.

JLockerman committed:
Release 1.1.1 on master, prepare for 1.2.0-dev

mfreed committed:
Update README to clarify relationship of Apache vs. TSL code

Clarify that code isn't dual licensed, but source code is either Apache-2 or
Timescale-Licensed, and you can build an Apache-2 or Timescale-Licensed binary.

Addresses question raised in https://github.com/timescale/timescaledb/issues/922

gayyappan committed:
Fix for deadlock between select and drop_chunks

Acquire foreign key related table locks prior to dropping chunks.  This
helps avoid deadlocks in scenarios where select and drop chunks are
trying to acquire locks on the same set of tables.
Changes to isolation test suite to prevent diffs related to displaying
hypertable id differences across test suite runs.

Fixes #865

amytai committed:
Move enterprise updates to newest update file

svenklemm committed:
Mark gapfill functions parallel safe

Gapfill functions need to be marked parallel safe to not prevent
parallelism. The gapfill node itself is still parallel restricted
but child nodes can be parallel

JLockerman committed:
Add reorder command

New cluster-like command which writes to a new index than swaps,
much like is done for the data table, and only acquires
exclusive locks for said swap. This trades off disk usage for
lower contention: we hold locks for a much lower period of time,
allowing reads to work concurrently, but we have both the old
and new versions of the table existing at once, approximately
doubling storage usage while reorder is running.

Currently only works on chunks.

svenklemm committed:
Improve CI settings

run pginstallcheck with parallel schedule
don't stop on dependency error for installcheck
don't build postgres with --enable-coverage

erimatnor committed:
Make header format consistent across TSL

This change makes include guards consistently formatted across TSL
code and also adds a missing include guard in `planner.h`. The include
guards include `TSL` in the prefix so that they don't conflict with
the ones in the same-named non-TSL header files.

JLockerman committed:
Add telemetry report test

We should really test that this function works, and we're outputting all
the fields we expect. Not testing the actual values as that changes
across installs and runs.

Should also improve our code coverage.

JLockerman committed:
Add .codecov.yml and ignore test/src

We don't care about the test coverage of our test infrastrucutre

dianasaur323 committed:
Make older_than / newer_than error message user friendly

dianasaur323 committed:
Make older_than / newer_than error message user friendly

JLockerman committed:
Edit launcher to load TSL strictly after timescale itself

Reloading shared libraries (which is done to start parallel
workers) is done is the order in which load_file completes.
If we try to load the TSL from PG_init, this causes postgres to
attempt to load the TSL before timescale itself, which causes an
error.

Also disables the warning on normal TSL startup.

JLockerman committed:
Update telemetry test for new telemetry

svenklemm committed:
Replace hardcoded database name from regression tests

Replace hardcoded database name from regression tests with :TEST_DBNAME
Remove creation of database single_2 from test runner and add it to
bgw_launcher and loader test since no other tests used those
use SQL comments in test scripts

JLockerman committed:
Stop using the extra field for now and other Windows bugs

Something is causing a heap corruption upon setting the license key to
default when we try to use the guc extra on windows. For now stop using
it and just rerun the validation function, if we get to the assign hook
we must have a valid key, so it will never fail.

Also Fixes error message on windows;
turns out windows does not like to print NULL strings.
Don't do that.

Fixes other minor windows bugs.

JLockerman committed:
Use a fixed file for the latest dev version

When developing a feature across releases, timescaledb updates can get
stuck in the wrong update script, breaking the update process. To avoid
this, we introduce a new file "latest-dev.sql" in which all new updates
should go. During a release, this file gets renamed to
"<previous version>--<current version>.sql" ensuring that all new
updates are released and all updates in other branches will
automatically get redirected to the next update script.

svenklemm committed:
Execute codecov for successful builds only

Execute codecov for successful builds only
Don't run codecov for pgindent and license check

svenklemm committed:
Fix bgw_policy test

mfreed committed:
Add new top-level licensing information

Clarify better how source code is split between Apache 2.0 and Timescale
License, to now serve as a reference file to which GitHub will point.  Also
cleanup NOTICE to point to LICENSE files.

svenklemm committed:
Fix time column handling in get_create_command

Fixes get_create_command to produce a valid create_hypertable
call when the column name is a keyword
Add test.execute_sql helper function to test support functions

svenklemm committed:
Fix typo in pginstallcheck output filename

svenklemm committed:
Create directory for postgres tablespace test

This is will not fix the tablespace test for remote instances
but local instances will work

svenklemm committed:
Fix update test script

Remove EXIT trap handler because it seems to occasionally capture exit
code from something that is not an update test resulting in error
state even though all update tests were successful.

svenklemm committed:
Fix typos in comments

JLockerman committed:
Reduce lock timeout on deadlock_dropchunkcs_select test

Detects the deadlock due to the lock timeout instead of the test runner
timeout. Speeds up the test.

JLockerman committed:
Move 1.2.0 updates to the correct file

JLockerman committed:
Fix casing in telemetry

Make the license.edition field for an Apache only
license "apache_only" to be consistent with
"enterprise" and "community". Request from solar.

JLockerman committed:
Update license header

Only have the copyright in the NOTICE. Hopefully
only having to update one place each year will
keep it consistent.

JLockerman committed:
Clean windows warnings

After this, the only remaining warnings on Windows are from postgres
itself.

svenklemm committed:
Add update test for PostgreSQL 10

Add an update test for updating the extension on PostgreSQL 10.
Change PostgreSQL 9.6 update test to only run triggered by cron.

JLockerman committed:
Add telemetry reports for timescaledb.last_tuned and last_tuned_version

Our tool timescaledb-tune records when it was last run in the
postgresql.conf file, which we can use in telemetry to find out how many
people are tuning their DB with that tool.
It may also record which version of the tool was run, so we're adding
the ability to report this now.

RobAtticus committed:
Remove out-dated and unused scripts

Nothing appears to use scripts/start-test-docker.sh, so it can be
safely removed. docker-run.sh was still referenced on our docs
page, but should soon be removed. It no longer actually provides
anything useful (i.e., any memory settings it used to have have
been removed) and is out-of-date with newer Docker setups. To
remove the burden of keeping it up to date, we should just remove.

cevian committed:
Fix pgtest schedule of tests

The way that ignores were used previously failed with parallel tests
because it only matched the first test in the parallel list of tests.

JLockerman committed:
Add dummy values to test last_tuned telemetry

Adds dummy values so that last_tuned telemetry gets
reported in our tests. Also moves the location of
the telemetry adding code to prevent a segfault

svenklemm committed:
Add time_bucket_gapfill tests for int2/4/8

The int2 and int8 versions of time_bucket_gapfill weren't well
tested since almost all the int tests were testing int4.

JLockerman committed:
Copy tsl into build directory in smoketest

We need to ensure we have all the source in order
to build timescaledb. Lacking this breaks the test.

RobAtticus committed:
Remove git merge/rebase lines from TSL

It seems that a merge conflict was not completely cleaned up after
in the Timescale License.

svenklemm committed:
Add include for compat.h to gapfill/interpolate.c

When compiling against Postgres 9.6 before 9.6.5 TupleDescAttr
is not defined, compat.h defines the macro for versions before 9.6.5

svenklemm committed:
Fix race condition in drop_rename_hypertable test

The size of the bgw_job_stat table depends on running jobs and might be
inconsistent between runs.

svenklemm committed:
Replace hardcoded database name in test files

Replace hardcoded databases name with :TEST_DBNAME in test scripts
and :DBNAME in include files. This will allow parallel execution
of test scripts.

svenklemm committed:
Execute regression tests in parallel

Change the schedule generation for the regression tests to generate
a parallel schedule to improve test runtime. Also moving tablespace
directory creation to pg_regress.sh to prevent redundant work and
race conditions between parallel runs.

svenklemm committed:
Fix order in timestamp test

The order specified in some of the timestamp tests did not result
in consistent ordering when a different plan was chosen.

svenklemm committed:
Differentiate between required license in error hints

Show different error hint depending on the license required
to use a feature.

JLockerman committed:
Add cron job to test ApacheOnly builds nightly on travis

It’s good to check if these work, and we definitely need to before
every release

JLockerman committed:
Reload license manually in BGW

Currently BGW jobs get guc values but do not necessarily run the assign
hook. This is an issue for the TSL because the TSL is loaded in just
such a hook; if they are never called, TSL functions cannot be used.
Fix by manually reloading the guc in the TSL.

JLockerman committed:
Fix signal processing in background workers

When receiving a SIGHUP, background workers should reload pg_conf to get
any changed guc values.

Also fix issue where the test bgw_scheduler was not properly waiting for
child processes to die before shutting itself down. The test scheduler
did not setup the sigterm handler, so it wouldn't shutdown child
processes correctly. This meant that our tests, which would later test
the number of jobs running, would depend on the exact timing signals
would propagte to bgw processes. This should increase the robustness of
bgw_db_scheduler.

Change tune telemetry guc to PGC_SIGHUP because it looks like BACKEND
gucs interact oddly with ALTER_SYSTEM in our tests.

JLockerman committed:
Allow module shutdown to work in ApacheOnly builds

In cases where we're going from ApacheOnly to ApacheOnly
allow a nop shutdown hook. This prevents spurious errors.

JLockerman committed:
Fix Apache only builds for older cmake versions

add_compile_definitions is to new to run on travis.

JLockerman committed:
Ignore the generated plan_gapfill tsl tests

These tests should never be committed to version control
as they need to be generated based on the version of postgres
installed locally.

JLockerman committed:
Disable parallel workers in background workers

We currently do not have a parallel worker context in out BGW
and we wouldn't want our BGW to start parallel workers anyway
in order to conserve resources. This commit disables parallel
workers in BGW jobs.

JLockerman committed:
Fix warnings for TSL licenses

So as to reduce the amount of logspam users receive, restrict printing license info
to the following:

  1. On CREATE EXTENSION
       a. in the notice, print the license expiration time, if any
       b. if the license is expired additionally print that
       c. else if the license will expire within a week print an addional warning
  2. On the first usage of a TSL function, print if the license is expired or will
     be expired within a week

JLockerman committed:
Clean error handling for invalid licenses

We shouldn't segfault

davidkohn88 committed:
Log to pg log when running reorder jobs

Log before and after we run the reorder job on a chunk so that there
is more information in the server logs around when and how reorder
jobs occur.

JLockerman committed:
Clean missing braces warning

This shouldn't be needed, but it fixes a spurious warning for those of
us that enable more warnings.

Commit:
Explicit chunk exclusion

In some cases user might already know what chunks need to be scanned to answer
a particular query. Using `chunks_in` function we can skip calculating chunks
involved in particular query which should result in better performances as well.
A simple example:

`SELECT * FROM hypertable WHERE chunks_in(hypertable, ARRAY[1,2])`

JLockerman committed:
Log which chunks are dropped by background workers

We don't want to do this silently, so that users are
able to debug where their chunks went.

JLockerman committed:
Set the reorder job to run immediately if there are more chunks in need of reordering

This will allow the job to catchup if there's a backlog of chunks in need of reordering.

JLockerman committed:
Fix speeling

JLockerman committed:
Add view displaying info about the current license

Currently the view displays the current edition, expiry date, and
whether the license is expired. We're not displaying the license key
itself in the view as it can get rather long, and get be read via SHOW.
We also do not display the license's ID since that is for internal use.

svenklemm committed:
Add ordered append optimization

This optimization will replace the MergeAppendPath for queries on
hypertables ordered by the time partitioning column and with a
LIMIT clause with an ordered AppendPath. This optimization will
remove the need for last point queries to access every chunk of
a hypertable.

This commit also adds struct TimescaleDBPrivate which is stored
in RelOptInfo->fdw_private to store TimescaleDB-specific plan state
between different planner hook invocations in the plan.
We needed to add TimescaleDBPrivate to store a flag indicating
whether or not to use ordered append between different parts of
the planner.

svenklemm committed:
Ignore rolenames pgregress test

The rolenames test seems to be fail quite often in CI, so we ignore
the result from that test for now.

svenklemm committed:
Remove year from SQL file license text

This changes the license text for SQL files to be identical
with the license text for C files.

JLockerman committed:
Use job_stat last started time on reorder fast restart instead real time

In reorder bgw fast restart, use the time we declared the reorder job
should start in the job_stat table instead of using the transaction's
start time as the next start time. This allows us to test that fast
restart operates correctly when running using a virtual timer.

Use the default schedule interval for integer open dimensions in reorder

This makes reorder jobs behave properly on non-time open dimensions

JLockerman committed:
Disallow scheduled drop_chunks policies on non-time open dimensions

Scheduled drop_chunks works by dropping chunks older than a certain
interval. Since we have no way to calculate what 'now()' is for non-time
dimensions, we currently cannot have a scheduled drop_chunks job that
works with them. Currently this causes an error when the policy attempts
to execute. This commit moves the error, so that it occurs when a user
tries to add the policy in the first place.

JLockerman committed:
Add tests of scheduled jobs within actual background workers

To date all our bgw tests run in the normal psql environment. This
commit adds tests that the jobs function correctly when run from actual
background workers. This should catch cases where the background worker
behaves differently than a normal connection.

JLockerman committed:
Only use seqscans for reorder

In our workloads we expect that there will never be a case in which an
indexscan based reorder will perform better than a seqscan based one.
In order to prevent pessimization, disable indexscan based reordering
for btrees, and only use it when we absolutely must.

svenklemm committed:
Run all diff tests after normal tests

plan_hashagg_results_x_diff and plan_expand_hypertable_results_diff
rely on output of previous tests so run them after other tests

JLockerman committed:
Point the travis build badge at "history" instead of "current"

Right now, clicking the travis build badge directs the user to the
currently running travis job. Since this is usually a PR, and we gate PR
merges on travis anyway, it's not very useful for determining the health
of the project. This commit switches the link to point at the build
history, which shows the previously run cron jobs; the jobs that the
badge points at anyway.

svenklemm committed:
Install findutils in build container

The find syntax used by the codecov script is not compatible with
busybox find, installing GNU find makes the codecov script functional
again.

cevian committed:
Fix bugs in chunk_scan

svenklemm committed:
Fix typo in background worker error message

JLockerman committed:
Fix bgw_launcher restart when failing during launcher setup

We do cleanup for bgw_message_queue in the before_shmem_exit function,
but we were registering said function after initializing the database
hashtable. When bgw_launcher would crash in middle of populating the
hashtable, this would cause issues as the bgw_message_queue wouldn't be
reset, and the launcher would fail to restart as it would think that
another launcher was running. This commit moves the hook registering to
before we touch any of the state that the hook cleans up, in order to
ensure that the state will be cleaned regardless of where the failure
occurs. This is already safe for the bgw_message_queue since that
function validates that the caller owns the queue before altering it,
and is made safe for the hashtable via an additional level of
indirection, so we can check if the hashtable exists.

This commit also cleans up some of the lock ordering around the
bgw_message_queue to prevent potential race conditions; specifically, if
we're doing a test-and-set, both the "test" and the "set" have to be
within the same lock-acquisition.

JLockerman committed:
Update telemetry with prettier os info

The info gotten from uname is difficult to work with, so read the os
name from /etc/os-release if it's available.

JLockerman committed:
Fix formatting for error message

dianasaur323 committed:
Update README for small typo

Small typo -> changing "build" to "built" when discussing Apache-ONLY

cevian committed:
Mark Postgresql includes in cmake as SYSTEM includes.

This is needed for tools like include-what-you-use to decide
which files should be included with quotes vs carrots.
This is probably more correct anyway as postgres files are
included with carrots in our code.

Also made the change for OpenSSL because it seems more correct.

svenklemm committed:
Use hypertable for parallel tests

Generated data has been offset by 1 to align with chunk boundaries.

svenklemm committed:
Remove variable declaration in for loops

Since we add PostgreSQL CFLAGS to our own CFLAGS on some platforms
using variable declarations in for loops will generate compile
warnings, depending on the CFLAGS PostgreSQL is compiled with.

Commit:
Making chunks_in function internal

This function needs chunk ids as input. Since chunk ids are
 TimescaleDB internal metadata it feels more natural to make this function internal.

svenklemm committed:
Bail on ordered append when pathkeys don't match

When an index is not available on all chunks pathkeys of the child
might not match pathkeys of the MergeAppendPath, PostgreSQL fixes
this when creating the merge append plan by inserting a sort node
for the child. Unfortunately this is too late for us so we don't do
this optimization for those cases for now.

svenklemm committed:
Change time_bucket_gapfill argument names

time_bucket_gapfill used end as argument name which is a sql keyword
and has to be quoted when used, this changes the argument names from
start/end to start/finish.

davidkohn88 committed:
Rename alter_policy_schedule & main_table for better UI

Rename alter_policy_schedule to alter_job_schedule for consistency with the job_id argument passed in.
Also rename main_table to hypertable in all of the policy related functions as they must deal with
hypertables that have already been created.

davidkohn88 committed:
Add informational views for policies

Add views so that users can see what the parameters are for policies they have created
and a separate view so that they can see policies that have been created and scheduled on hypertables.

JLockerman committed:
Release 1.2.0

JLockerman committed:
Fix issues with non-dev versions when generating the update scripts

Without this, we generate multiple rules for the latest script

svenklemm committed:
Make time_bucket_gapfill start and finish optional

Make time_bucket_gapfill start and finish optional, this is in
preparation for deducing them from WHERE clause. We make this
optional now to not introduce breaking change later. This also
only allows simple expressions for bucket_width, start and
finish because only those can be evaluated safely in gapfill_begin.

cevian committed:
Fix for ExecSlotDescriptor during upserts

Previously we set the mt_existing and onConflictProjInfo slot
descriptors only when the current chunk had a tup_conv_map this
was based on a false assumption that these slot descriptors had the
hypertable's slot descriptor. In fact, since we are reusing these slots
across chunks, the slot descriptor will correspond to the previous
chunk's one. This led to an error if the previous chunk had a
tup_conv_map but the current one did not.

Most of the work done here was by @gayyappan with minor improvements by
@cevian. The latter is submitting because of a time-crunch having to do
with upcoming release.

cevian committed:
Make slot for upserts be update for every chunk switch

Previously, the slots for upserts were only updated when
a ChunkInsertState was created, but not when it was found in the
ChunkDispatch cache. This was wrong -- we need to update the slots
every time we switch to the chunk.

Right now the switch function is called on every tuple. We will
optimize that later to only be called when the previous chunk was
different. But, that is not part of this bug-fix PR.

Ngalstyan4 committed:
Fix some DirectFunctionCall calls (correctly convert arguments to Datum)

Ngalstyan4 committed:
Change long to int64

Depending on host hardware, long can be 32 or 64 bits long.
Changing it to int64 or int32 makes the code more portable and helps
avoid issues when running in 32bit systems.

cevian committed:
Make a copy of the ri_onConflict object in PG11

In PG11 the OnConflictProjInfo object is a member of ri_onConflict.
Previously, the ri_onConflict object was copied by reference when
creating our rri. But since OnConflictProjInfo has to change for each
chunk and CIS this was incorrect as a change to OnConflictProjInfo
changed the "global" ri_onConflict.

This commit changes the copy of ri_onConflict to a deep one so that each
CIS now has it's own independent OnConflictProjInfo.

Ngalstyan4 committed:
Fix Datum conversion in gapfill

gayyappan committed:
switch cis state when we switch chunks.

JLockerman committed:
Prepare the repo for the next development cycle

JLockerman committed:
Fix loader for old versions of timescale

The loader needs to be backwards compatible. Therefore we handle the case where post_load function doesn't
exist gracefully.

JLockerman committed:
Add warning about restarting loader for this release

And a "thanks" to the user who found the bug that requires it

JLockerman committed:
Fix loading the tsl license in parallel workers

Parallel workers have an unusual load path, which doesn't behave the
same as other processes. Since we don't have any obvious hooks we can
use to enable loading, and the library load order is already frozen
by the main process, fix this by adding a _PG_init which will enable
the tsl in background workers when needed.

erimatnor committed:
Make code formatting cover TSL test source code

This change makes code formatting (`pgindent`) cover the TSL test
sources under `tsl/test/src/`. It also addresses some minor formatting
issues with existing sources in other directories.

erimatnor committed:
Fix release build tests and exclude debug/test code

This change excludes test and debug code from the release build of the
TSL library and also makes tests properly work under release
builds. Excluding test/debug code is important since otherwise a
significant amount of dead code will inflate the size of the library.

Some issues with the TSL isolation tests have also been
addressed. Generated spec files are removed from the GitHub repository
and are now output to the build/binary directory instead. To be able
to run specs from the build directory, the `pg_isolation_regress.sh`
script has been updated to be able to run tests from a non-default
specs dir.

JLockerman committed:
Run travis ci with optimizations turned on

Our users will pretty much always run timescale with optimizations, so
we might as well run our CI in a more realistic setup. Plus, this may
make CI run faster, since build times are negligible.

JLockerman committed:
Add cron a pre-release Release mode testing

Most of our testing is performed with Debug enabled. This is desired
as it enables more assertions, and helps ensure that our code is
correct. Still, it's good to ensure we're not accidentally depending
on Debug code in Release builds, so run tests in that mode nightly
and before each release.

JLockerman committed:
Fix typo failing our update 9.6 update tests

svenklemm committed:
Stop background worker on rename database

When background workers are running RENAME DATABASE will not work
because there are still active sessions to that database open.
This stops background workers on RENAME DATABASE so databases can
be renamed without manually stopping background workers.

svenklemm committed:
Check for NULL pointer in ts_bookend_finalfunc

When the sfunc is never called to initialize InternalCmpAggStore
state can be NULL even though PG_ARGISNULL is false, so we check for
that condition here. Also add a similar check to ts_hist_finalfunc.

svenklemm committed:
Fix typos in comments

svenklemm committed:
Fix group handling in time_bucket_gapfill

When time_bucket_gapfill is used with multiple groups
but the query returned no rows we can not generate any
rows in the gapfill node.

cevian committed:
Run clang-format on code
